New mysql impl [review]

ljacqu opened the pull request:
local pr to start a review

ljacqu commented on pom.xml:
```diff
@@ -189,6 +189,43 @@
+                    <groupId>org.slf4j</groupId>
+                </exclusion>
+            </exclusions>
+            <optional>true</optional>
```
Is HikariCP provided by Bukkit or so? Not sure why `<optional>true</optional>` here and below

EbonJaeger replied:
I don't think so, but that's also how it is in AuthMe. Knowing me, it's probably something else I'm ignorant of. lol

ljacqu commented on src/main/java/me/gnat008/perworldinventory/PerWorldInventory.java:
```diff
@@ -112,7 +111,7 @@ public void onEnable() {
         injector.register(Server.class, getServer());
         injector.register(PluginManager.class, getServer().getPluginManager());
-        injector.provide(DataFolder.class, getDataFolder());
+        injector.provide(DataFolder.class, new File(getDataFolder(), "data"));
```
Good :heavy_check_mark:

ljacqu commented on src/main/java/me/gnat008/perworldinventory/PerWorldInventory.java:
```diff
@@ -151,8 +150,8 @@ public void onDisable() {
-        serializer = injector.getSingleton(FileWriter.class);
-        injector.register(DataWriter.class, serializer);
+        serializer = injector.getSingleton(DataSource.class);
+        injector.register(DataSource.class, serializer);
```
This is no longer necessary since it's the same class. If you don't use the serializer field anywhere else here, you can delete these two lines and the serializer field.

EbonJaeger replied:
As below, yeah, I kinda forgot about that.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/DataSourceProvider.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Creates the data source
+ */
+public class DataSourceProvider implements Provider<DataSource> {
```
For this to take effect it needs to be registered as the provider of DataSource. Somewhere in the main class' onEnable process: `injector.registerProvider(DataSource.class, DataSourceProvider.class)` (or similar; typing this off the top of my head)

EbonJaeger replied:
Ah yeah, I saw that. Forgot to actually do it. hehe

ljacqu commented on src/main/java/me/gnat008/perworldinventory/config/DatabaseProperties.java:
```diff
@@ -0,0 +1,68 @@
+public final class DatabaseProperties implements SettingsHolder {
+
+    @Comment({"Type of database to use",
+        "Valid values: flatfile, sqlite, mysql"})
```
as long as sqlite isn't implemented and throws an exception I wouldn't list it in the values

EbonJaeger replied:
True. I'm in a weird place of figuring it will be done with the MySQL release, but I just don't have everything for it yet.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/FlatFile.java:
```diff
@@ -41,21 +24,19 @@
-    private final PWIPlayerFactory pwiPlayerFactory;
+    private PerWorldInventory plugin;
+    private PlayerSerializer playerSerializer;
+    private PWIPlayerFactory pwiPlayerFactory;
```
These fields are no longer set anywhere. They need to be either passed to the constructor as before or you need to make the data source provider have all of those classes so it can pass them along. If you want to do the first one (my recommendation), you can `@Inject Injector injector` into the data source provider and then call the appropriate class.

The advantage of a provider class is you can perform other stuff during the initialization process. In AuthMe, that's wrapping the data source with a cachedDataSource if needed, and checking the SQLite size to issue a warning in some cases. If you want to do something like that in the future, keep with the provider and either inject all required fields, or inject the injector. (You generally want to avoid injecting the injector, but for small contained classes like a provider, I'd argue that it's fine.)
If you don't plan on doing any additional work on the initialization of the data source, you could map each data source type to a class (add a field on the DataSourceType enum) and then just get the type from the settings, get the class and call the injector.

EbonJaeger replied:
Yeah this is the problem I ended on last night. It needs all those, but I couldn't figure out a nice way to get them to it. Since it's now being created by the provider, I can't just `@Inject` them like i did before.

So you're saying to inject the injector in the provider class, and have that pass all the dependencies to the flatfile via its constructor?

ljacqu replied:
Two possibilities :)
```java
public enum DataSourceType {
   FLAT_FILE(FlatFile.class),
   MYSQL(MySql.class);
   //...
   public Class<? extends DataSource> getDataSourceClass() { /* ... */ }
}

public class PerWorldInventory {
   public void onEnable() {
      // ...
      DataSource dataSource = injector.getSingleton(
           settings.getProperty(BACKEND).getDataSourceClass());
      injector.register(DataSource.class, dataSource);
   }
}
```

Or, what you've asked:
```java
public class DataSourceProvider extends Provider<DataSource> {
   @Inject 
   private Injector injector;

   @Override 
   public DataSource get() {
       switch (settings.getProperty(BACKEND)) {
           case FLAT_FILE:
              return injector.getSingleton(FlatFile.class);
           case MYSQL:
               return injector.getSingleton(MySql.class);
           default:
               throw new IllegalArgumentException("Unknown data source type");
      }
   }
}
```

EbonJaeger replied:
And then I would have the fields or constructor in FlatFile have the `@Inject` annotation? (With the second option)

ljacqu replied:
Yes

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/MySQL.java:
```diff
@@ -0,0 +1,288 @@
+     * @param ex The raised exception.
+     */
+    private static void logSQLException(SQLException ex) {
+        PwiLogger.severe("Error during SQL operation:" + ex);
```
for this and 2x below: unless you really want this, use `PwiLogger.severe("Error ...", ex)` instead

EbonJaeger replied:
Good point. Not sure why I didn't just do that...

ljacqu commented on src/test/java/me/gnat008/perworldinventory/TestHelper.java:
```diff
@@ -16,6 +19,7 @@
 public final class TestHelper {
 
+    public static final String SOURCES_FOLDER = "src/main/java/";
     public static final String PROJECT_ROOT = "/me/gnat008/perworldinventory/";
```
in the listener consistency test, these constants can be used also.

ljacqu commented on src/test/java/me/gnat008/perworldinventory/ReflectionTestUtils.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Offers reflection functionality to set up tests. Use only when absolutely necessary.
+ */
+public final class ReflectionTestUtils {
```
TestHelper also has some reflection methods, please move/merge those into this new class

EbonJaeger replied:
Noted, will do.

ljacqu commented on src/test/java/me/gnat008/perworldinventory/config/SettingsConsistencyTest.java:
```diff
@@ -17,37 +25,110 @@
+    }
+
+    /**
+     * Make sure that all properties in the config contain a default value.
```
> a default value

the _same_ default value as in the code

ljacqu commented on src/test/java/me/gnat008/perworldinventory/config/SettingsConsistencyTest.java:
```diff
@@ -17,37 +25,110 @@
     }
 
+    /**
+     * Make sure that all properties defined in both the config and the code.
```
This especially checks that config.yml does not have any properties that are unknown in the code. The above method checks the other way around: that config.yml contains all properties as defined in the code.

ljacqu commented on src/test/java/me/gnat008/perworldinventory/config/SettingsConsistencyTest.java:
```diff
@@ -17,37 +25,110 @@
+                    String fieldName = "Field " + clazz.getSimpleName() + "#" + field.getName();
+                    assertThat(fieldName + " should be public, static and final",
+                            isValidConstantField(field), equalTo(true));
+                }
```
good :heavy_check_mark:

ljacqu commented on src/test/java/me/gnat008/perworldinventory/config/SettingsConsistencyTest.java:
```diff
@@ -17,37 +25,110 @@
+
+        System.out.println("Found " + classes.size() + " SettingsHolder implementations");
+
+        configData = SettingsRetriever.buildConfigurationData();
```
You just scanned for all SettingsHolder classes -- use them to build your configData. `ConfigurationDataBuilder.collectData(classes)`

ljacqu commented on src/main/java/me/gnat008/perworldinventory/commands/SetWorldDefaultCommand.java:
```diff
@@ -15,7 +15,7 @@
 
     @Inject
-    private FileWriter fileSerializer;
+    private FlatFile fileSerializer;
```
If you have operations that will always manipulate files (regardless of the settings) you should extract those methods into their separate class. It can get confusing fast otherwise as to why a concrete data source type is injected at some places. I'd wager that the methods used in this class are NOT part of the dataSource interface anyway.

Besides an issue of clarity it's important to be aware that if you inject a FlatFile directly, you'll get ANOTHER instance of FlatFile than if you inject DataSource (provided that it's mapped to FlatFile). This may bite you in the ass later on as it's hard to track down :P

EbonJaeger replied:
That's actually a really good idea, I hadn't thought of that. I had so far just figured "oh it handles files, put all the methods in there!", but you're right. Especially since they need to be used as well for the setWorldDefaults command.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/config/SettingsRetriever.java:
```diff
@@ -0,0 +1,28 @@
+                PwiProperties.class,
+                DatabaseProperties.class
+        );
+    }
```
This is OK but you need to decide which way to go here. You currently use settings holder classes in three different ways:
1. provided directly to Settings
1. this new class wrapping what you do in Settings
1. class collector in settings consistency test

Especially 1 and 2 need to be merged together. Advantage of 2. is that you only define your settings holder classes at one point and after that you can reuse it everywhere. In AuthMe this makes sense because we have many tests where we want the ConfigurationData of AuthMe's settings.
3 is OK and I'd recommend to leave it because this way you really get _all_ settings holder classes. If you remove a class from this one here, and you remove all of the class' properties in config.yml, the consistency test will still fail because you forgot to delete the class.
What would probably be even better (and I know we don't do it in Authme, either), is to have this class list here live as a field on the class and to have the settings consistency test check that he didn't find more classes with the class collector.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/MySQL.java:
```diff
@@ -0,0 +1,288 @@
+                    "FOREIGN KEY (pid) REFERENCES " + prefix + "players(pid));";
+            statement.executeUpdate(sql);
+
+            PwiLogger.info("MySQL tables created!");
```
Nice code :heavy_check_mark: 

One thing to look into: are backticks supported by all DB types? I think this is a MySQL thing only.
`PwiLogger.info("MySQL tables created!");` will be logged each time the plugin starts :)

EbonJaeger replied:
As I recall, they are not supported by all types. I've never really actually used SQL databases in code, so there is still a lot I'm trying to figure out. I think the last time I tried to do this though, it was the only way I got things to work right.

Also, oops. I seem to recall I made that same mistake last time... hehe.

Gnat008 commented on src/test/java/me/gnat008/perworldinventory/commands/SetWorldDefaultCommandTest.java:
```diff
@@ -34,6 +35,8 @@
 /**
  * Tests for {@link SetWorldDefaultCommand}.
  */
+// TODO Gnat008: Fix tests so they work again
```
I think all the tests for this one need to be reworked, and I'm struggling with how to do it currently.

EbonJaeger replied:
I think all the tests for this one need to be reworked, and I'm struggling with how to do it currently.

EbonJaeger commented on src/test/java/me/gnat008/perworldinventory/config/SettingsConsistencyTest.java:
```diff
@@ -107,7 +109,7 @@ public void shouldHavePublicStaticFinalFields() {
+        for (Class<? extends SettingsHolder> clazz : classes) {
             Field[] fields = clazz.getDeclaredFields();
             for (Field field : fields) {
                 if (Property.class.isAssignableFrom(field.getType())) {
```
Cant add a comment to the line below where this actually happens, but I'm getting this:
http://hastebin.com/zihawixomi.vbs

ljacqu commented:
@Gnat008 
I addressed some points in a commit on this fork—https://github.com/ljacqu/PerWorldInventory/commit/fcc73c6f3af3ccd03f23e5fcaea7a1f3a4650a8b.

- I criticized your use of SettingsRetriever.buildConfiguration() right after having collected all settings holder classes. While fixing the test I realized it's quite a pain to turn a `List<Class<? extends SettingsHolder>>` into an equivalent array. Sorry about that. 

- Fixed the compiling issue with calling ReflectionTestUtils#getField by creating a second method, `ReflectionTestUtils.getField(Object instance, Field field)` since we already have the field when we make the call there. No sense in passing only the field name and then getting the field _again_.

- So you don't wonder: I changed the TemporaryRule field from private to public in FileUtilsTest because it's a requirement by JUnit. Test class fails otherwise.

- SetWorldDefaultCommandTest:
   - Added some initial changes to make the tests working again. Mockito's runner doesn't support custom annotations (such as `@DataFolder`), so if you find yourself injecting stuff like that you have to switch to the `DelayedInjectionRunner`. It will search the test class for a similarly annotated field, and `@BeforeInjecting` allows to do some setup before the `@InjectDelayed` fields are initialized.
  - You moved methods from FlatFile into the SetWorldDefaultCommand class. For testing, this leaves you with two choices: 1) a.) you either need to change your verification to check that certain files have been written, or b.) you can make the methods package-private, make a spy out of the command object you test and then you can still call verify(command).blabla(). IMHO not a nice approach because a.) is cumbersome to implement and b.) gets confusing fast as to which methods are _really_ being called and which methods are being mocked on the spy. I'd suggest a second approach: 2) moving file writing methods to their own class. This way you can inject a mock into the command class again.

ljacqu commented:
You can either pick the changes from that commit manually, or on command line, if you have _your_ new-mysql-impl branch checked out with no local changes:

```
git remote add ljacqu https://github.com/ljacqu/PerWorldInventory.git
git merge ljacqu/new-mysql-impl
git reset origin/new-mysql-impl
git remote remove ljacqu
```

Leaves you with my commit as local changes :)

ljacqu commented:
Summary of our call:
- refactoring needed to only interact with DataSource interface so you don't have to know whether flat file or MySQL is configured
- you want to keep the number of queries as low as possible, the most expensive operation is most likely to be setting up a connection to the DB
- might make sense to reduce the number of tables since insert statements have to be done by table individually, and update statements might be complicated
- might be worth to look into having a unique index on the (uuid, gamemode, group) tuple
- primary key on "secondary" tables might be interesting for some DB performance reason, otherwise they could be dropped (needs research)
